### PR TITLE
replicated and derived tables with separate serializers

### PR DESCRIPTION
**Pull request by olgavrou**
Regarding issue #539 addressing action points 1 and 2.
Regarding issue #562 addressing two middle check points (runtime consensus switch, passing switch to enclave/nodestate)

- `Map` has a boolean field `replicated` that indicates if the map is a replicated or a derived one. By default if we are using Raft this will be set to `true`. If we are in PBFT it will be set to `false`, so any PBFT replicated tables need to be explicitly defined as replicated ones
- We expose this property via the `TxView` so that `Tx::serialise()` knows if the tx view over a map it serialising is replicated or derived
- `Tx::serialise()` returns a `SerialisedMaps` struct that contains 2 byte-vectors, one for replicated maps and one for derived maps, instead of returning a single byte-vector
- `Tx::serialise()` creates two serializers, one for each type of table and returns the raw data from each. Each serializer continues to serialize the data based on the domain (i.e. first the data from the public tables and then the private ones)
- Modified `GenericSerialiseWrapper`  to check that there are actual data written in the serializer, and if not return an empty vector. This is needed since in the constructor we pass in the version to be serialised. Since now we create two serializers, one can be empty if there are no replicated or derived maps. However because the serializer has been created, `get_raw_data()` would return a vector with size 1, since we have written the version on construction
- We continue to pass only the replicated maps when `replicate()` is called in `Store::commit()` but history now gets both the replicated and derived write sets and can decide what to do with them. For now we just concatenate the two vectors and use them to update the merkle tree. The other option would be to have `SerialisedMaps` contain a vector over replicated maps and a vector over all-maps (instead of replicated and derived), and then have the merkle tree updated with the all-maps vector. But then we either serialize and encrypt the replicated maps twice or have to find a workaround to serialize replicated maps only once but reference them twice
- kv unit tests updated to explicitly declare the tables as replicated in order for them to pass when building with PBFT

### CODE REVIEW

File: src/node/history.h [hunk 1]
```diff
@@ -398,8 +399,14 @@ namespace ccf
     void add_result(
       kv::TxHistory::RequestID id,
       kv::Version version,
-      const std::vector<uint8_t>& data) override
+      const std::vector<uint8_t>& data_replicated,
+      const std::vector<uint8_t>& data_derived) override
     {
+      std::vector<uint8_t> data;
+      data.reserve(data_replicated.size() + data_derived.size());
```
Comment (jumaffre) [inline src/node/history.h]: Minor: This could just be `std::vector<uint8_t> data(data_replicate.size() + data_derived.size());`

File: src/kv/test/kv_test.cpp [hunk 2]
```diff
@@ -745,4 +760,74 @@ TEST_CASE("private recovery map swap")
       // impossible during recovery.
     }
   }
+}
+
+TEST_CASE("replicated and derived table serialisation")
```
Comment (jumaffre) [inline src/kv/test/kv_test.cpp]: Could this be moved to `kv_serialisation.cpp` which is the unit test file for KV serialisation (also automatically run as part of `kv_test`).

File: src/kv/test/kv_test.cpp [hunk 3]
```diff
@@ -745,4 +760,74 @@ TEST_CASE("private recovery map swap")
...
+    REQUIRE(serialised.derived.size() == 0);
+  }
+
+  {
+    Store::Tx tx(store.next_version());
+    auto data_view = tx.get_view(data_derived);
+    auto data_view_private = tx.get_view(data_derived_private);
+    data_view->put(42, 42);
+    data_view_private->put(43, 43);
+    auto [success, reqid, serialised] = tx.commit_reserved();
```
Comment (jumaffre) [inline src/kv/test/kv_test.cpp]: Should we try to use `commit_reserved()` as little as possible here? You could get the serialised data by attaching the `StubConsensus` to the store.

File: src/kv/test/kv_test.cpp [hunk 4]
```diff
@@ -745,4 +760,74 @@ TEST_CASE("private recovery map swap")
...
+    REQUIRE(
+      store.deserialise(serialised.derived) == kv::DeserialiseSuccess::PASS);
+
+    REQUIRE(serialised.derived.size() > 0);
+    REQUIRE(serialised.replicated.size() == 0);
+  }
+
+  {
+    Store::Tx tx(store.next_version());
+    auto data_view_r = tx.get_view(data_replicated);
```
Comment (jumaffre) [inline src/kv/test/kv_test.cpp]: Minor: you can get all the views in one `get_view()`.

File: src/kv/test/kv_test.cpp (hunk 3 above)
Comment (olgavrou) [reply src/kv/test/kv_test.cpp]: I do not get why. `commit_reserved` will call `tx::serialise()` which is what I want to test here, there is no consensus involved.

File: src/kv/kv.h [hunk 5]
```diff
@@ -995,8 +1013,8 @@ namespace kv
 
       version = c.value();
 
-      const std::vector<uint8_t> data = serialise();
-      if (!data.size())
+      SerialisedMaps data = serialise();
```
Comment (jumaffre) [inline src/kv/kv.h]: `const auto`?

File: src/kv/kv.h [hunk 6]
```diff
@@ -995,8 +1013,8 @@ namespace kv
...
+      if (data.replicated.empty() && data.derived.empty())
```
Comment (jumaffre) [inline src/kv/kv.h]: Minor: You could define a `empty()` for `SerialisedMaps`.

File: src/kv/kv.h (hunk 5 above)
Comment (olgavrou) [reply src/kv/kv.h]: `auto` yes, `const` no since we are doing a `move` on the `data` a few lines down and if it is `const` it will invoke the copy ctor instead of the move one.

File: src/kv/kv.h [hunk 7]
```diff
@@ -1184,6 +1210,11 @@ namespace kv
     template <class K, class V, class H = std::hash<K>>
     using Map = Map<K, V, H, S, D>;
     using Tx = Tx<S, D>;
+#ifdef PBFT
```
Comment (jumaffre) [inline src/kv/kv.h]: I think the `KV` should not know about PBFT. Could the default behaviour be set in `networktables.h` instead?

Comment (olgavrou) [reply src/kv/kv.h]: So this is a good question. The thing is that we do not only create tables in `networktables.h`. There are a lot of app tables to, which means we will have to check if we are using PBFT or not in a lot of places. It is not ideal for the KV to know about PBFT but this is the place where this change has the minimal effect (afaik)

File: src/kv/genericserialisewrapper.h [hunk 8]
```diff
@@ -196,10 +196,26 @@ namespace kv
       // writer.
       if (!crypto_util)
       {
+        if (serialised_public_domain.size() <= 1)
```
Comment (jumaffre) [inline src/kv/genericserialisewrapper.h]: At the moment, the constructor of `GenericSerialiseWrapper` serialises the version but it would be easy to either 1) serialise the version after the `GenericSerialiseWrapper` is created (we should enforce that the version should be serialised before anything else) or 2) add a state enum to detect whether only the version has been serialised. I think 2 is neater and more extendable. 

Then, you can `return {}` based on that enum.

Comment (olgavrou) [reply src/kv/genericserialisewrapper.h]: Thanks I will think about this again. I looked into it a little bit. There is the minor caveat that we also have a `remove` function in `GenericSerialiseWrapper` so you would need to make checks on `remove` too in order to make sure that you still have data that you want to serialise, which makes it a bit less neater. You will still need to check if only the version is left in the end.

Comment (eddyashton) [reply src/kv/genericserialisewrapper.h]: These removes aren't removing things from the serialised data, they're serialising removals (a kind of write?) from the kv. I think the minimal enum solution here is to add a `bool ignore = false` arg to `serialise_internal`. When we call `serialise_internal(version_)` from the constructor, we can ignore this and don't change our `has_contents` enum, but any _other_ calls to `serialise_internal` will try to set the enum.

File: src/node/history.h (hunk 1 above)
Comment (olgavrou) [reply src/node/history.h]: Not in this case. If we set the size on initialization we get a vector that size but with default-initialized values populating the vector, where as if we use `reserve` it just reserves the memory but doesn't initialize it. This makes a difference since we are using `data.end()` when we do `data.insert()`.

File: src/kv/test/kv_test.cpp (hunk 2 above)
Comment (olgavrou) [reply src/kv/test/kv_test.cpp]: Done :)

File: src/kv/genericserialisewrapper.h (hunk 8 above)
Comment (olgavrou) [reply src/kv/genericserialisewrapper.h]: Implemented with a `bool` arg on `serialise_internal` and a `bool` `has_contents` field

**Comment by cimetrics** [comment]
images

File: src/kv/kv.h (hunk 7 above)
Comment (jumaffre) [reply src/kv/kv.h]: As discussed, one of the options would be to wrap the `Store` object that is available to any frontend (`tables`) and in `networktables.h` in an object that is responsible for creating the maps. Then, we can use this object to define the `replicated_default` for maps when they are created.

File: src/kv/genericserialisewrapper.h [hunk 9]
```diff
@@ -76,22 +76,31 @@ namespace kv
...
 
     std::shared_ptr<AbstractTxEncryptor> crypto_util;
 
     // must only be set by set_current_domain, since it affects current_writer
     SecurityDomain current_domain;
 
     template <typename T>
-    void serialise_internal(T&& t)
+    void serialise_internal(T&& t, bool content_update = true)
     {
```
Comment (achamayou) [inline src/kv/genericserialisewrapper.h]: Adding a branch on every single serialise_internal is unfortunate, is there another way to do this?

**Comment by cimetrics** [comment]
images

File: src/kv/kv.h (hunk 7 above)
Comment (jumaffre) [reply src/kv/kv.h]: As discussed, this is OK for now.

Comment (achamayou) [reply src/kv/kv.h]: Why does this need to be static? Why can't we set that on the KV at construction time? We want to move away from ifdef and towards a runtime switch for the consensus.

Comment (olgavrou) [reply src/kv/kv.h]: As discussed, will modify this so that the consensus can pass either none, all, or a list of tables that it wants to replicate. Then the KV can set this bool for each table on creation.

File: src/kv/genericserialisewrapper.h (hunk 9 above)
Comment (olgavrou) [reply src/kv/genericserialisewrapper.h]: As discussed I will either go back to checking the size of the serializer for `1` or remove the changes from `GenericSerialiseWrapper` all together and have the `serialise()` function decide if each serializer has actual data or not.

**Comment by olgavrou** [comment]
As discussed with @achamayou and @ashamis, the PR will be updated so that the derived write set is also passed through to `replicate()`. That way when we do catchup in PBFT, or when we are replaying the ledger, we don't need to execute the requests.

**Comment by cimetrics** [comment]
images

File: src/enclave/enclave.h [hunk 10]
```diff
@@ -97,7 +101,7 @@ namespace enclave
       start_type = start_type_;
       ccf_config = ccf_config_;
 
-      auto r = node.create({start_type, ccf_config_});
+      auto r = node.create({start_type, consensus_type, ccf_config});
```
Comment (olgavrou) [inline src/enclave/enclave.h]: passing to nodestate even though it is not used in this PR

**Comment by cimetrics** [comment]
images

File: src/kv/kv.h [hunk 11]
```diff
@@ -1627,15 +1683,22 @@ namespace kv
           if (success_ != CommitSuccess::OK)
             LOG_DEBUG_FMT("Failed Tx commit {}", last_replicated + offset);
 
+          std::vector<uint8_t> all_data;
```
Comment (olgavrou) [inline src/kv/kv.h]: concatenating `replicated` and `derived` write sets here instead of inside `history` as we are passing all of the serialised data to `replicate()`

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: src/host/main.cpp [hunk 12]
```diff
@@ -320,6 +325,14 @@ int main(int argc, char** argv)
                                   public_rpc_address.hostname,
                                   node_address.port,
                                   rpc_address.port};
+  if (consensus == "raft")
+  {
+    consensus_type = ConsensusType::Raft;
```
Comment (achamayou) [inline src/host/main.cpp]: CLI11 1.8 has nicer support for enums (see https://github.com/CLIUtils/CLI11/pull/233), this is separate from this PR, but it would be nice to upgrade and use that support (or at least to ticket it for now :)).

File: src/kv/kv.h [hunk 13]
```diff
@@ -1142,18 +1146,22 @@ namespace kv
...
             curr_map->serialise(derived_serialiser, include_reads);
           }
         }
       }
 
       // Return serialised Tx.
-      return {std::move(replicated_serialiser.get_raw_data()),
-              std::move(derived_serialiser.get_raw_data())};
+      return {replicated ? std::move(replicated_serialiser.get_raw_data()) :
+                           std::move(std::vector<uint8_t>(0)),
```
Comment (achamayou) [inline src/kv/kv.h]: {}

File: src/kv/kvtypes.h [hunk 14]
```diff
@@ -51,6 +51,13 @@ namespace kv
     PASS_SIGNATURE = 2
   };
 
+  enum ReplicateType
+  {
+    ALL = 0,
+    NONE,
+    SPECIFIED
```
Comment (achamayou) [inline src/kv/kvtypes.h]: SOME? :)

**Comment by achamayou** [comment]
@olgavrou not sure if I'm reading too much into that, but it looks like the last few cimetrics agree on a fairly consistent -5% on KV serialise. It's not enough that I'm worried about it, but it may be interesting to do some local kv_bench runs to check if that's accurate or not.

File: src/kv/kv.h (hunk 13 above)
Comment (olgavrou) [reply src/kv/kv.h]: afaik you can't use braced initializer list on a ternary operator

**Comment by olgavrou** [comment]
@achamayou regarding the metrics running the bench mark for some time (master vs my PR) I get the below results (roughly):

image

File: cmake/common.cmake [hunk 15]
```diff
@@ -558,17 +558,16 @@ set_property(TARGET http_parser.host PROPERTY POSITION_INDEPENDENT_CODE ON)
...
-  unset(PBFT_ARG)
+  set(CONSENSUS_ARG "raft")
 endif()
 
 set(CCF_NETWORK_TEST_ARGS
   ${TEST_IGNORE_QUOTE}
   ${TEST_ENCLAVE_TYPE}
   -l ${TEST_HOST_LOGGING_LEVEL}
   -g ${CCF_DIR}/src/runtime_config/gov.lua
-  ${PBFT_ARG}
```
Comment (jumaffre) [inline cmake/common.cmake]: Why was this moved out and duplicated?

**Comment by cimetrics** [comment]
images

File: src/host/main.cpp (hunk 12 above)
Comment (olgavrou) [reply src/host/main.cpp]: https://github.com/microsoft/CCF/issues/569

File: src/kv/kv.h (hunk 13 above)
Comment (achamayou) [reply src/kv/kv.h]: Indeed, I didn’t know that!

**Comment by achamayou** [comment]
@olgavrou thanks for checking, that looks like it’s within noise range.

**Comment by cimetrics** [comment]
images

File: cmake/common.cmake (hunk 15 above)
Comment (olgavrou) [reply cmake/common.cmake]: It's not the same argument. I am passing `--consensus pbft/raft` instead of `--pbft`. But I did move it back into `CCF_NETWORK_TEST_ARGS`

**Comment by cimetrics** [comment]
images
